pkg/command/context_tkg_helper.go

```diff
@@ -0,0 +1,192 @@
+	// installed plugin binary path(cli.PluginInfo.InstallationPath)
+	sOut, sErr, err := tkfo.CmdExecutor(rootCmd, args)
+	if err != nil {
+		return sErr, errors.Wrap(err, string(sErr))
+	}
+	return sOut, nil
+}
+
+func (tkfo *TKGKubeconfigFetcherOptions) isPinnipedPluginInstalled() (bool, error) {
+	standalonePlugins, err := pluginsupplier.GetInstalledStandalonePlugins()
+	if err != nil {
+		return false, err
+	}
+
+	for i := range standalonePlugins {
+		if standalonePlugins[i].Name == pinnipedAuthPluginName {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (tkfo *TKGKubeconfigFetcherOptions) getKubeconfigForPriorTKGVersion() (string, string, error) {
+	var kubeCfg, kubeCtx string
+	// While this would add an extra HTTP round trip, it avoids the need to
+	// add extra provider specific login flags.
+	tlsConfig, err := tkgauth.GetTLSConfig(tkfo.EndpointCACertPath, tkfo.InsecureSkipTLSVerify)
+	if err != nil {
+		return "", "", errors.Wrap(err, "failed to get TLS config")
+	}
+
+	isVSphereSupervisor, err := wcpauth.IsVSphereSupervisor(endpoint, getDiscoveryHTTPClient(tlsConfig))
+	// Fall back to assuming non vSphere supervisor.
+	if err != nil {
+		err := fmt.Errorf("error creating kubeconfig with tanzu pinniped-auth login plugin: %v", err)
+		log.Error(err, "")
+		return "", "", err
+	}
+
+	if isVSphereSupervisor {
+		log.Info("Detected a vSphere Supervisor being used")
+		kubeCfg, kubeCtx, err = vSphereSupervisorLogin(endpoint)
+		if err != nil {
+			err := fmt.Errorf("error logging in to vSphere Supervisor: %v", err)
+			log.Error(err, "")
+			return "", "", err
+		}
+	} else {
+		kubeCfg, kubeCtx, err = tkgauth.KubeconfigWithPinnipedAuthLoginPlugin(endpoint, nil,
+			tkgauth.DiscoveryStrategy{ClusterInfoConfigMap: tkgauth.DefaultClusterInfoConfigMap}, tkfo.EndpointCACertPath, tkfo.InsecureSkipTLSVerify)
+		if err != nil {
+			err := fmt.Errorf("error creating kubeconfig with tanzu pinniped-auth login plugin: %v", err)
+			log.Error(err, "")
+			return "", "", err
+		}
+	}
+	return kubeCfg, kubeCtx, nil
+}
+
+// RunCommandAndGetStdOutAndErr executes the cobra command and returns the command's stdout,stderr and error
+func RunCommandAndGetStdOutAndErr(cmd *cobra.Command, args []string) ([]byte, []byte, error) {
+	// Create a pipe to read stdout
+	ro, wo, err := os.Pipe()
+	if err != nil {
+		return nil, nil, err
+	}
+	c := make(chan []byte)
+	go fetchOutput(ro, c)
+
+	// Create a pipe to read stderr
+	re, we, err := os.Pipe()
+	if err != nil {
+		return nil, nil, err
+	}
+	e := make(chan []byte)
+	go fetchOutput(re, e)
+
+	stdout := os.Stdout
+	stderr := os.Stderr
+	defer func() {
+		os.Stdout = stdout
+		os.Stderr = stderr
+	}()
+	os.Stdout = wo
+	os.Stderr = we
+	cmd.SetArgs(args)
+	err = cmd.Execute()
+	wo.Close()
+	we.Close()
+	sout := <-c
+	serr := <-e
+	if err != nil {
+		return sout, serr, err
+	}
+	return sout, serr, nil
```

Comment:
Is it possible to simplify this a bit?
```
func RunCommandAndGetStdOutAndErr(cmd *cobra.Command, args []string) ([]byte, []byte, error) {
	var stderr bytes.Buffer
	var stdout bytes.Buffer
	cmd.SetArgs(args)
	cmd.SetOut(&stdout)
	cmd.SetErr(&stderr)
	err := cmd.Execute()
	if err != nil {
		return stdout.Bytes(), stderr.Bytes(), err
	}
	return stdout.Bytes(), stderr.Bytes(), nil
}
```

Reply:
No. It doesn't work unless the cobra command implementation(plugin) uses the output buffer(or any writer) set for the command. Eg:  [here](https://github.com/vmware-tanzu/tanzu-cli/blob/7d455e5ac56612db028b22d8fd755b9d1d9bbe58/pkg/command/plugin_search.go#L67). I asked pinniped-auth team for the same for stdout, however we need the same for capturing stderrr( command implementer should have used cmd.ErrOrStderr() instead of writing directly to os.stderr.  Inorder to cater for all the scenarios, this approach was taken. 
However I have added `TODO` comment [here](https://github.com/vmware-tanzu/tanzu-cli/pull/309/files#diff-b7561fb8167076584a78310ef324f8434967a2607d1e17a5b592bd546645895aR88), where we can explore using the  pluginInfo.InstalledPath and then use the binary path in exec.Command(). We can evaluate the option later and update the logic if we have merits

Reply:
Based on the discussion with Prem, it seems like the way the `pinniped-auth` plugin writes output and error, this approach might not be possible to do because of the following reason:
-  To use this approach command should be writing stdout and stderr using `cmd.OutOrStdOut` and `cmd.ErrOrStdErr` which is not happening with the existing and new pinniped-auth plugin which writes directly to stdout and stderr.

Reply:
When we call a plugin from the CLI, we don't control where the output should go.
We call a plugin binary here: https://github.com/vmware-tanzu/tanzu-cli/blob/a6778fb0adaa5ac28f205d7e20a015be2b1a6238/pkg/cli/plugin_cmd.go#L26

Notice that in that function we have the `cmd` argument which is the `tanzu` cli.  We may be able to use the output streams of that cmd (using `cmd.OutOrStderr()` and `cmd.OutOrStdout()`) to request the plugin output to be redirected.